src/mpol/coordinates.py

```diff
@@ -50,109 +51,208 @@ class GridCoords:
+        return self._f_edges
 
-        # for evaluating a packed vis... uu, vv increasing + fftshifted
-        self.packed_u_centers_2D = np_fft.fftshift(self.sky_u_centers_2D)
-        self.packed_v_centers_2D = np_fft.fftshift(self.sky_v_centers_2D)
+    @property
+    def v_edges(self) -> NDArray[floating[Any]]:
+        return self._f_edges
 
-        # and in polar coordinates too
-        self.packed_q_centers_2D = np_fft.fftshift(self.sky_q_centers_2D)
-        self.packed_phi_centers_2D = np_fft.fftshift(self.sky_phi_centers_2D)
+    @property
+    def u_centers(self) -> NDArray[floating[Any]]:
+        return self._f_centers
 
-        self.q_max = float(
-            np.max(np.abs(self.packed_q_centers_2D)) + np.sqrt(2) * self.du
-        )  # outer edge [klambda]
+    @property
+    def v_centers(self) -> NDArray[floating[Any]]:
+        return self._f_centers
 
-        # x_centers_2D and y_centers_2D are just l and m in units of arcsec
-        x_centers_2D, y_centers_2D = np.meshgrid(
-            self.l_centers / const.arcsec, self.m_centers / const.arcsec, indexing="xy"
-        )  # [arcsec] cartesian indexing (default)
+    @property
+    def u_bin_min(self) -> float:
+        return self._min_f
 
-        # for evaluating a packed cube... ll, mm increasing + fftshifted
-        self.packed_x_centers_2D = np_fft.fftshift(x_centers_2D)  # [arcsec]
-        self.packed_y_centers_2D = np_fft.fftshift(y_centers_2D)  # [arcsec]
+    @property
+    def v_bin_min(self) -> float:
+        return self._min_f
 
-        # for evaluating a sky image... ll mirrored, mm increasing, no fftshift
-        self.sky_y_centers_2D = y_centers_2D  # [arcsec]
-        self.sky_x_centers_2D = np.fliplr(x_centers_2D)  # [arcsec]
+    @property
+    def u_bin_max(self) -> float:
+        return self._max_f
 
-    def check_data_fit(self, uu: npt.ArrayLike, vv: npt.ArrayLike) -> None:
+    @property
+    def v_bin_max(self) -> float:
+        return self._max_f
+
+    @property
+    def img_ext(self) -> list[float]:
+        # calculate the image extent
+        # say we had 10 pixels representing centers -5, -4, -3, ...
+        # it should go from -5.5 to +4.5
+        lmax = self.cell_size * (self.npix // 2 - 0.5)
+        lmin = -self.cell_size * (self.npix // 2 + 0.5)
+        return [lmax, lmin, lmin, lmax]  # arcsecs
+
+    @property
+    def vis_ext(self) -> list[float]:
+        return [
+            self.u_bin_min,
+            self.u_bin_max,
+            self.v_bin_min,
+            self.v_bin_max,
+        ]  # [kλ]
+
+    # --------------------------------------------------------------------------
+    # Non-identical u & v properties
+    # --------------------------------------------------------------------------
+    @cached_property
+    def sky_u_centers_2D(self) -> NDArray[floating[Any]]:
+        # only useful for plotting a sky_vis
+        # uu increasing, no fftshift
+        return np.tile(self.u_centers, (self.ncell_u, 1))
```

Comment:
What is the reason for choosing `tile` over `meshgrid`?

Reply:
`meshgrid(x, y)` returns tiled `x` and `y.T` at the same time. With the properties, I'm defining only one of the attributes so used `tile` instead. If you think `meshgrid` feels more natural (and readable), I can pull the definitions of these attributes up to the constructor and define it that way.

Reply:
OK I added a code comment so that it's more clear what tile is doing. I'm used to seeing meshgrid many places in the numpy codebase, so that was obvious to me. But I agree `tile` makes more sense here.